fabtests/pingpong: EFA competition traces and a race condition fix

alekswn opened the pull request:
These changes should make AWS CI failures more informative and fix occasional client-side timeouts in 32 byte pingpong with efa-direct.

- **EFA RX/TX completions**
Currently we have traces for posting send and receive buffers with EFA provider. This change adds completion traces as well.

~~- **Unexpected completion check in `ft_finalize()`**~~
~~We have seen issues caused by duplicate EFA completions, adding check for unexpected ones to `ft_finalize()`.~~

~~- **Use `ft_sync()` instead of `ft_finalize_ep(ep)` if `ft_finalize()`**.~~
~~- **Call ft_sync() before ft_finalize()**~~
- **Follow `ft_sync_inband(false)` pattern in `ft_finalize()`**

There are two effects of this change:
1. If out-of-band sync was enabled it's going to be used for the final sync as well.
2. In case of inband sync server will wait for client to process all recv completions before it can exit.

alekswn commented on fabtests/common/shared.c:
```diff
@@ -3261,7 +3299,11 @@ int ft_finalize(void)
+	if (ret)
+		return ret;
+
+	return ft_check_for_unexpected_comps();
```
Right `ft_sync()` posts an another one RX buffer. Added call to `ft_finalize_ep()`

a-szegel commented on fabtests/benchmarks/benchmark_shared.c:
```diff
@@ -239,6 +238,12 @@ int pingpong(void)
 		if (ret)
 			return ret;
+
+		/* Final inband sync requires preposted RX buffers */
```
It seems wrong that you are adding this in one benchmark `pingpong()`, and modifying its use in common code which gets called in every benchmark. `ft_finalize()` -> `ft_check_for_unexpected_comps()`

shijin-aws replied:
There should be an outstanding rx ahead in `ft_enable_ep_recv`, do we really need this additional rx here?

alekswn replied:
I have an impression we're consuming all RX-es at `pingpong_no_pre_posted_rx()` here:
https://github.com/ofiwg/libfabric/blob/99a50e7c2f9afd352ec3c74af66dfca35be23126/fabtests/benchmarks/benchmark_shared.c#L166-L167
https://github.com/ofiwg/libfabric/blob/99a50e7c2f9afd352ec3c74af66dfca35be23126/fabtests/benchmarks/benchmark_shared.c#L179-L180

I contrast to `pingpong_pre_posted_rx()` which reposts RX buffer in `ft_rx()`:
https://github.com/ofiwg/libfabric/blob/99a50e7c2f9afd352ec3c74af66dfca35be23126/fabtests/common/shared.c#L2588-L2610

alekswn commented:
> It's also failing in efa provider
> 
Need to fix logic in `ft_check_for_unexpected_comps`

alekswn commented:
In the new revision:
- fixed logic in ` ft_check_for_unexpected_comps()`
- moved `ft_sync()` call to the top of `ft_finalize()`
- deleted redundant `ft_sync()` calls from tests calling it explicitly

shijin-aws commented on fabtests/functional/multi_mr.c:
```diff
@@ -269,7 +269,6 @@ static int run_test(void)
 
 	ret = mr_key_test();
 
-	ft_sync();
```
Can we leave ft_sync and `ft_finalize` separate? You just need to add a missing `ft_sync` in pingpong right?

alekswn replied:
Sure, will do that

shijin-aws commented on fabtests/common/shared.c:
```diff
@@ -2907,6 +2907,44 @@ int ft_get_tx_comp(uint64_t total)
 	return ret;
 }
 
+static inline int ft_check_for_unexpected_comps()
```
The motivation for this change is detecting whether we really have a duplicate completion, this is for experimental purpose so I do not think we want this detection in the production code.

alekswn replied:
Ok, will delete it

alekswn commented:
bot:aws:retest

j-xiong reviewed (commented):
The changes to the pingpong test look fine. I would expect some description in the commit message about the reason behind the change.

shijin-aws commented on fabtests/benchmarks/benchmark_shared.c:
```diff
@@ -283,6 +283,13 @@ int run_pingpong(void)
 			return ret;
 	}
 
+	/* ft_finalize() needs RX buffer to be posted for proper sync */
```
Should this be moved to `pingpong_no_pre_posted_rx` ?

shijin-aws replied:
@iziemba can you take a look at this change ?

alekswn replied:
Not that easy to move it to `pingpong_no_pre_posted_rx()`. The function is called from `pingpong()`. Pingpong is called from two places above. One of them in a loop. We need that at the end of the last iteration only.

shijin-aws replied:
Ah, u are right.

alekswn commented:
Changes in the new revision:
- rewritten `ft_finalize_ep()` to follow `ft_sync_inband()` structure

alekswn commented:
bot:aws:retest

shijin-aws commented on fabtests/benchmarks/benchmark_shared.c:
```diff
@@ -283,6 +283,13 @@ int run_pingpong(void)
 
+	/* ft_finalize() needs RX buffer to be posted for proper sync */
+	if (ft_check_opts(FT_OPT_NO_PRE_POSTED_RX)) {
+		ret = ft_post_rx(ep, rx_size, &rx_ctx);
```
efa-direct requires a pre-posted rx buffer otherwise the sender will get stuck at RNR

shijin-aws commented:
bot:aws:retest

aingerson commented on fabtests/common/shared.c:
```diff
@@ -3236,17 +3236,25 @@ int ft_finalize_ep(struct fid_ep *ep)
-		return ret;
+	if (opts.dst_addr) {
+		ret = ft_tx_msg(ep, remote_fi_addr, tx_buf, 4, &ctx,
+				FI_TRANSMIT_COMPLETE);
```
If you update or rerun the CI can you fix the alignment of this line and then also line 3254. Thanks!

aingerson reviewed (approved):
Approving but there are two misaligned lines which I'm pointing out since it's your first PR but I'm not going to be annoying about it if you don't need to retest. But if you're planning on retriggering the CI anyway, please fix.
